File: app-java/src/main/java/com/tritiumstudios/phasmophobiaevidencetool/android/firebase/firestore/transactions/user/account/transaction/FirestoreUnlockHistory.java
package com.tritiumstudios.phasmophobiaevidencetool.android.firebase.firestore.transactions.user.account.transaction;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.tritiumstudios.phasmophobiaevidencetool.android.firebase.firestore.listeners.OnFirestoreProcessListener;
import com.tritiumstudios.phasmophobiaevidencetool.android.firebase.firestore.transactions.user.account.FirestoreTransactionHistory;
import com.google.firebase.Timestamp;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.SetOptions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class FirestoreUnlockHistory {

    private final static String COLLECTION_UNLOCK_HISTORY = "UnlockHistory";

    private final static String DOCUMENT_UNLOCKED_ITEM = "UnlockedItem";

    private final static String FIELD_DATE_UNLOCKED = "dateUnlocked";
    private final static String FIELD_TYPE = "type";

    public static void init()
            throws Exception {

        getUnlockedDocument().get()
                .addOnSuccessListener(unused ->
                        Log.d("Firestore", COLLECTION_UNLOCK_HISTORY + " successfully INITIALIZED!"))
                .addOnFailureListener(e -> {
                    Log.d("Firestore", COLLECTION_UNLOCK_HISTORY + "User Unlock History failed INITIALIZATION");
                    e.printStackTrace();
                })
                .addOnCompleteListener(task ->
                        Log.d("Firestore", COLLECTION_UNLOCK_HISTORY + "User Unlock History INITIALIZATION process complete!"));

    }

    @NonNull
    public static CollectionReference getUnlockHistoryCollection()
            throws Exception {

        return FirestoreTransactionHistory.getTransactionHistoryDocument().collection(COLLECTION_UNLOCK_HISTORY);
    }

    @NonNull
    private static DocumentReference getUnlockedDocument()
            throws Exception {
        CollectionReference purchaseHistoryCollection = getUnlockHistoryCollection();

        return purchaseHistoryCollection.document(DOCUMENT_UNLOCKED_ITEM);
    }

    public static void addUnlockDocument(
            @Nullable String unlockUUID, String type, @NonNull OnFirestoreProcessListener callback)
            throws Exception {
        CollectionReference unlocksCollection = getUnlockHistoryCollection();

        if(unlockUUID == null) { return; }

        Map<String, Object> documentData = new HashMap<>();
        documentData.put(FIELD_TYPE, type);
        documentData.put(FIELD_DATE_UNLOCKED, Timestamp.now());

        DocumentReference unlockDocument = unlocksCollection.document(unlockUUID);

        unlockDocument.set(documentData, SetOptions.merge())
                .addOnSuccessListener(unused -> {
                    callback.onSuccess();

                    Log.d("Firestore", "Unlocked document of " + unlockUUID + " GENERATED / LOCATED!");
                })
                .addOnFailureListener(e -> {
                    callback.onFailure();

                    Log.d("Firestore", "Unlocked document of " + unlockUUID + " could NOT be GENERATED / LOCATED!");
                    e.printStackTrace();
                })
                .addOnCompleteListener(task -> {
                    callback.onComplete();

                    Log.d("Firestore", "Unlocked document of " + unlockUUID + " process complete.");
                });
    }

    public static void addUnlockedDocuments(
            @Nullable ArrayList<String> unlockUUIDs, String type, @NonNull OnFirestoreProcessListener callback)
            throws Exception {
        CollectionReference unlockedHistoryCollection = getUnlockHistoryCollection();

        if(unlockUUIDs == null) { return; }

        Map<String, Object> documentData = new HashMap<>();
        documentData.put(FIELD_TYPE, type);
        documentData.put(FIELD_DATE_UNLOCKED, Timestamp.now());

        for(String uuid : unlockUUIDs) {
            DocumentReference purchasedDocument = unlockedHistoryCollection.document(uuid);
            purchasedDocument.set(documentData, SetOptions.merge())
                    .addOnSuccessListener(unused -> {
                        callback.onSuccess();

                        Log.d("Firestore", "Unlocked document of " + uuid + " GENERATED / LOCATED!");
                    })
                    .addOnFailureListener(e -> {
                        callback.onFailure();

                        Log.d("Firestore", "Unlocked document of " + uuid + " could NOT be GENERATED / LOCATED!");
                        e.printStackTrace();
                    })
                    .addOnCompleteListener(task -> {
                        callback.onComplete();

                        Log.d("Firestore", "Unlocked document of " + uuid + " process complete.");
                    });
        }
    }
}

File: app-compose/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/data/firebase/firestore/transactions/user/account/transactions/types/FirestoreUnlockHistory.kt
package com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.account.transactions.types

import android.util.Log
import com.google.android.gms.tasks.Task
import com.google.firebase.Timestamp
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.SetOptions
import com.tritiumgaming.phasmophobiaevidencepicker.firebase.firestore.transactions.user.account.transactions.FirestoreTransactionHistory.Companion.transactionHistoryDocument
import com.tritiumgaming.phasmophobiaevidencepicker.data.listeners.firestore.OnFirestoreProcessListener

class FirestoreUnlockHistory {

    companion object {
        private const val COLLECTION_UNLOCK_HISTORY = "UnlockHistory"

        private const val DOCUMENT_UNLOCKED_ITEM = "UnlockedItem"

        private const val FIELD_DATE_UNLOCKED = "dateUnlocked"
        private const val FIELD_TYPE = "type"

        @Throws(Exception::class)
        fun init() {
            unlockedDocument.get()
                .addOnSuccessListener { unused: DocumentSnapshot? ->
                    Log.d("Firestore",
                        "$COLLECTION_UNLOCK_HISTORY successfully INITIALIZED!")
                }
                .addOnFailureListener { e: Exception ->
                    Log.d("Firestore",
                        "$COLLECTION_UNLOCK_HISTORY User Unlock History failed INITIALIZATION")
                    e.printStackTrace()
                }
                .addOnCompleteListener { task: Task<DocumentSnapshot?>? ->
                    Log.d("Firestore",
                        "$COLLECTION_UNLOCK_HISTORY User Unlock History INITIALIZATION process complete!")
                }
        }

        val unlockHistoryCollection: CollectionReference
            @Throws(Exception::class)
            get() = transactionHistoryDocument.collection(COLLECTION_UNLOCK_HISTORY)

        private val unlockedDocument: DocumentReference
            @Throws(Exception::class) get() {
                val purchaseHistoryCollection = unlockHistoryCollection

                return purchaseHistoryCollection.document(DOCUMENT_UNLOCKED_ITEM)
            }

        @Throws(Exception::class)
        fun addUnlockDocument(
            unlockUUID: String?, type: String, callback: OnFirestoreProcessListener
        ) {
            if (unlockUUID == null) {
                return
            }

            val documentData: MutableMap<String, Any> = HashMap()
            documentData[FIELD_TYPE] = type
            documentData[FIELD_DATE_UNLOCKED] = Timestamp.now()

            val unlockDocument = unlockHistoryCollection.document(unlockUUID)

            unlockDocument.set(documentData, SetOptions.merge())
                .addOnSuccessListener {
                    callback.onSuccess()
                    Log.d("Firestore",
                        "Unlocked document of $unlockUUID GENERATED / LOCATED!")
                }
                .addOnFailureListener { e: Exception ->
                    callback.onFailure()
                    Log.d("Firestore",
                        "Unlocked document of $unlockUUID could NOT be GENERATED / LOCATED!")
                    e.printStackTrace()
                }
                .addOnCompleteListener {
                    callback.onComplete()
                    Log.d("Firestore",
                        "Unlocked document of $unlockUUID process complete.")
                }
        }

        @Throws(Exception::class)
        fun addUnlockedDocuments(
            unlockUUIDs: ArrayList<String>?, type: String, callback: OnFirestoreProcessListener
        ) {
            if (unlockUUIDs == null) {
                return
            }

            val documentData: MutableMap<String, Any> = HashMap()
            documentData[FIELD_TYPE] = type
            documentData[FIELD_DATE_UNLOCKED] = Timestamp.now()

            for (uuid in unlockUUIDs) {
                val purchasedDocument = unlockHistoryCollection.document(uuid)
                purchasedDocument.set(documentData, SetOptions.merge())
                    .addOnSuccessListener {
                        callback.onSuccess()
                        Log.d("Firestore",
                            "Unlocked document of $uuid GENERATED / LOCATED!")
                    }
                    .addOnFailureListener { e: Exception ->
                        callback.onFailure()
                        Log.d(
                            "Firestore",
                            "Unlocked document of $uuid could NOT be GENERATED / LOCATED!")
                        e.printStackTrace()
                    }
                    .addOnCompleteListener {
                        callback.onComplete()
                        Log.d("Firestore",
                            "Unlocked document of $uuid process complete.")
                    }
            }
        }
    }

}
